### PR TITLE
Storage.clone maintains original device

### LINKED ISSUE #14673: Deepcopy doesn't copy to same gpu
This issue was first reported here: https://discuss.pytorch.org/t/deepcopy-with-multiple-gpus/20283

```
from copy import deepcopy
import torch

x = torch.ones((1,), device=torch.device("cuda", 1))
print(x)
## result : tensor([ 1.], device='cuda:1') 
y = deepcopy(x)
print(y)
## result : tensor([ 1.], device='cuda:0') 
```
This code will copy x onto the first GPU with cuda id 0, instead of on the original GPU with cuda id 1.

deepcopy is necessary when we want to avoid autograd tracking gradients into the new copy, unlike .clone()

### PR DESCRIPTION
**Pull request by fmassa**
Fixes https://github.com/pytorch/pytorch/issues/14673

As pointed out by @vishwakftw , the root case of the `deepcopy` issue was that `storage.clone()` would create a new storage in the default device.

### CODE REVIEW

File: torch/storage.py [hunk 1]
```diff
@@ -39,7 +39,9 @@ def __sizeof__(self):
 
     def clone(self):
         """Returns a copy of this storage"""
-        return type(self)(self.size()).copy_(self)
+        device = self.get_device() if self.is_cuda else -1
+        with torch.cuda.device(device):
```
[the remark this answers is not in the document]
Comment (fmassa) [reply torch/storage.py]: This is the "old" way of setting devices in PyTorch, when devices were just numbers.
It is still used in a few places, [such as here](https://github.com/pytorch/pytorch/blob/master/torch/_utils.py#L68)

[the remark this answers is not in the document]
Comment (vishwakftw) [reply torch/storage.py]: Thank you for the reference.

**Comment by zou3519** [comment]
@fmassa -- most of the other functions in storage.py need this fix as well (`half`, `float`, `double`, etc...) :/

**Review by zou3519 — APPROVED** [review]
lgtm. It would be nice if we fixed this for the other functions in storage.py as well (it should be quick).

**Comment by fmassa** [comment]
@zou3519 it looks like this is already fixed (I'm not 100% sure how / why)
```python
a = torch.rand(2, device='cuda:1').storage()
print(a.double().get_device()  # 1
print(a.long().get_device()  # 1
print(a.half().get_device()  # 1
```

I can maybe add some more tests for this case?

**Comment by zou3519** [comment]
Oh. I'm really confused why those work but `.clone()` doesn't/didn't work.
Either way, feel free to ship this, I'd be curious to dig into this a little more now or later

File: torch/storage.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (SsnL) [reply torch/storage.py]: This is a bit of a hack. So is https://github.com/pytorch/pytorch/blob/ecc17fe3dd03962f36a989659336e42de86a38ca/torch/cuda/__init__.py#L503-L505. We should add proper ctors for Storage some day.
